experimental/lib/Transforms/Speculation/PlacementFinder.cpp

```diff
@@ -0,0 +1,359 @@
+// Note: srcBB and dstBB can be equal when the arcs are Backedges
+struct BlockPredecessor {
+  unsigned srcBB;
+  unsigned dstBB;
+  std::vector<OpPlacement> arcs;
+};
+using BBPredecessorMap = std::map<unsigned, std::vector<BlockPredecessor>>;
+
+// Get a map from BBs to the list of its BlockPredecessors
+static BBPredecessorMap getBlockPredecessors(handshake::FuncOp funcOp) {
+  // Traverse all operations to find arcs between BBs (and self-arcs)
+  BBEndpointsMap endpointArcs;
+  funcOp->walk([&](Operation *op) {
+    for (Value operand : op->getOperands()) {
+      BBEndpoints endpoints;
+      if (isBackedge(operand, op, &endpoints) or
+          endpoints.srcBB != endpoints.dstBB) {
+        OpPlacement arc = {operand, op};
+        endpointArcs[endpoints].push_back(arc);
+      }
+    }
+  });
+
+  // Join all predecessors of a BB
+  BBPredecessorMap predecessors;
+  for (const auto &[endpoints, arcs] : endpointArcs) {
+    BlockPredecessor predecessor;
+    predecessor.srcBB = endpoints.srcBB;
+    predecessor.dstBB = endpoints.dstBB;
+    predecessor.arcs = arcs;
+    predecessors[endpoints.dstBB].push_back(predecessor);
+  }
+
+  return predecessors;
+}
+
+// DFS traversal to mark all operations that lead to Commit units
+static void markSpeculativePaths(Operation *currOp,
+                                 SpeculationPlacements &placements,
+                                 PlacementList &markedPaths) {
+  for (Value res : currOp->getResults()) {
+    for (Operation *succOp : res.getUsers()) {
+      OpPlacement arc = {res, succOp};
+      if (not markedPaths.count(arc)) {
+        markedPaths.insert(arc);
+        // Stop traversal if a commit is reached
+        if (not placements.containsCommit(res, succOp))
+          markSpeculativePaths(succOp, placements, markedPaths);
+      }
+    }
+  }
+}
+
+void PlacementFinder::findCommitsBetweenBBs() {
+  OpPlacement specPos = placements.getSpeculatorPlacement();
+  auto funcOp = specPos.dstOp->getParentOfType<handshake::FuncOp>();
+  assert(funcOp && "op should have parent function");
+
+  // Whenever a BB has two speculative inputs, commit units are needed to avoid
+  // tokens going out-of-order. First, the block predecessor arcs are found
+  BBPredecessorMap predecessors = getBlockPredecessors(funcOp);
+  PlacementList markedPaths;
+  markSpeculativePaths(specPos.dstOp, placements, markedPaths);
+  // Iterate all BBs to check if commits are needed
+  for (const auto &[bb, blockPredecessors] : predecessors) {
+    // Count number of speculative inputs to the BB
+    unsigned countSpecInputs = 0;
+    for (BlockPredecessor pred : blockPredecessors) {
+      if (llvm::any_of(pred.arcs,
+                       [&](OpPlacement p) { return markedPaths.count(p); }))
+        countSpecInputs++;
+    }
+
+    if (countSpecInputs > 1) {
+      // Potential ordering issue, add commits
+      for (BlockPredecessor pred : blockPredecessors) {
+        for (OpPlacement arc : pred.arcs) {
```

Comment:
just because a bb has 2 spec arcs doesn't mean every arc needs a commit, and in fact will produce incorrect functioning. Think adding an if statement on if "arc" is in markedPaths would work?
Also don't think the individual edges are arcs.

Reply:
Thinking about this more I think it actually functions fine, its just a resource waste.

Reply:
This has been addressed by checking if the input edge is speculative or not before adding a commit.